ArduPlane/pullup.cpp

```diff
@@ -0,0 +1,244 @@
+        if (pitchup_complete && airspeed_low && !roll_control_lost) {
+                gcs().send_text(MAV_SEVERITY_INFO, "Pullup level r=%.1f p=%.1f alt %.1fm AMSL",
+                                ahrs.roll_sensor*0.01, ahrs.pitch_sensor*0.01, current_loc.alt*0.01);
+                break;
+        } else if (pitchup_complete && roll_control_lost) {
+                // push nose down and wait to get roll control back
+                gcs().send_text(MAV_SEVERITY_ALERT, "Pullup level roll bad r=%.1f p=%.1f",
+                                ahrs.roll_sensor*0.01,
+                                ahrs.pitch_sensor*0.01);
+                stage = Stage::PUSH_NOSE_DOWN;
+        }
+        return false;
+    }
+    case Stage::NONE:
+        break;
+    }
+
+    // all done
+    stage = Stage::NONE;
+    return true;
+}
+
+/*
+  stabilize during pullup from balloon drop
+ */
+void GliderPullup::stabilize_pullup(void)
+{
+    const float speed_scaler = plane.get_speed_scaler();
+    switch (stage) {
+    case Stage::WAIT_AIRSPEED: {
+        SRV_Channels::set_output_scaled(SRV_Channel::k_elevator, elev_offset*4500);
+        SRV_Channels::set_output_scaled(SRV_Channel::k_rudder, 0);
+        plane.nav_pitch_cd = 0;
+        plane.nav_roll_cd = 0;
+        uint32_t now = AP_HAL::millis();
+        if (now - plane.last_stabilize_ms > 1000) {
+            plane.pitchController.reset_I();
+            plane.yawController.reset_I();
+        }
+        plane.last_stabilize_ms = now;
+        SRV_Channels::set_output_scaled(SRV_Channel::k_aileron, plane.rollController.get_rate_out(0, speed_scaler));
+        plane.logger.Write_PID(LOG_PIDR_MSG, plane.rollController.get_pid_info());
+        ng_demand = 0.0;
+        break;
+    }
+    case Stage::WAIT_PITCH: {
+        plane.yawController.reset_I();
+        plane.nav_roll_cd = 0;
+        plane.nav_pitch_cd = 0;
+        plane.last_stabilize_ms = AP_HAL::millis();
+        SRV_Channels::set_output_scaled(SRV_Channel::k_rudder, 0);
+        SRV_Channels::set_output_scaled(SRV_Channel::k_aileron, plane.rollController.get_rate_out(0, speed_scaler));
+        float aspeed;
+        const auto &ahrs = plane.ahrs;
+        if (ahrs.airspeed_estimate(aspeed)) {
+            // apply a rate of change limit to the ng pullup demand
+            ng_demand += MAX(ng_jerk_limit / ahrs.get_EAS2TAS(), 0.1f) * plane.scheduler.get_loop_period_s();
+            ng_demand = MIN(ng_demand, ng_limit);
+            const float VTAS_ref = ahrs.get_EAS2TAS() * aspeed;
+            const float pullup_accel = ng_demand * GRAVITY_MSS;
+            const float demanded_rate_dps = degrees(pullup_accel / VTAS_ref);
+            const uint32_t elev_trim_offset_cd = 4500.0f * elev_offset * (1.0f - ng_demand / ng_limit);
+            SRV_Channels::set_output_scaled(SRV_Channel::k_elevator, elev_trim_offset_cd + plane.pitchController.get_rate_out(demanded_rate_dps, speed_scaler));
```

Comment:
Should we set `SRV_Channels::set_output_scaled(SRV_Channel::k_elevator, elev_trim_offset_cd)` if there is no airspeed? Currently it does not set in that case. That should mean its still at the value from the last loop, but its possible we set it somewhere else so that is not the case.

Reply:
i've added it, but in reality airspeed_estimate() always returns true if we have ever had a GPS lock, which will mean it's always true for this command